src/lib/OccAnalysis.hs

```diff
@@ -408,62 +410,50 @@ instance HasOCC (ComposeE PrimHof SAtom) where
       -- affecting that reference.  Using `IxAll` is a conservative
       -- approximation (in downstream analysis it means "assume I touch every
       -- value").
-      bd' <- oneShot2 a (Deterministic []) IxAll bd
+      bd' <- oneShot a [Deterministic [], IxAll]bd
       return $ RunState Nothing ini' bd'
     RunState (Just _) _ _ ->
       error "Expecting to do occurrence analysis before destination passing."
-    RunIO bd ->
-      -- I think IxAll is right, since indexing with a value of Unit type
-      -- does mean indexing with every possible value of the index type.
-      RunIO <$> oneShot a IxAll bd
+    RunIO bd -> RunIO <$> occ a bd
     RunInit _ ->
       -- Though this is probably not too hard to implement.  Presumably
       -- the lambda is one-shot.
       error "Expecting to do occurrence analysis before lowering."
-    CatchException _ ->
-      error "Expecting to do occurrence analysis after exception resolution."
-      -- But it's one-shot if we need it: CatchException <$> oneShot bd
-    Linearize _ -> error "Expecting to do occurrence analysis after AD."
-    Transpose _ -> error "Expecting to do occurrence analysis after AD."
     Seq _ _ _ _ ->
       error "Expecting to do occurrence analysis before destination passing."
     RememberDest _ _ ->
       error "Expecting to do occurrence analysis before destination passing."
 
--- Occurrence analysis of the body of a one-shot unary lambda Atom.
-oneShot :: Access n -> IxExpr n -> SAtom n -> OCCM n (SAtom n)
-oneShot acc binderSummary bd =
-  occLam bd \b' body -> extend b' (sink binderSummary) do
-    occ (sink acc) body
-
--- Occurrence analysis of the body of a one-shot binary lambda Atom.
-oneShot2 :: Access n -> IxExpr n -> IxExpr n -> SAtom n -> OCCM n (SAtom n)
-oneShot2 acc binderSummary1 binderSummary2 bd =
-  occLam bd \b' (AtomicBlock body)
-    -> AtomicBlock <$> do
-      extend b' (sink binderSummary1) do
-        oneShot (sink acc) (sink binderSummary2) body
+-- -- Occurrence analysis of the body of a one-shot unary lambda Atom.
+-- oneShot2 :: Access n -> IxExpr n -> IxExpr n -> LamExpr SimpIR n -> OCCM n (LamExpr SimpIR n)
+-- oneShot2 acc binderSummary1 binderSummary2 (BinaryLamExpr b1 b2 body) = do
+--   occWithBinder (Abs b1 (Abs b2 body)) \b1' (Abs b2' body') -> do
+--     extend b1' (sink binderSummary1) do
+--       UnaryLamExpr b2'' body'' <- oneShot (sink acc) (sink binderSummary2) (UnaryLamExpr b2' body')
+--       return $ BinaryLamExpr b1' b2'' body''
+
+oneShot :: Access n -> [IxExpr n] -> LamExpr SimpIR n -> OCCM n (LamExpr SimpIR n)
```

Comment:
> In particular, are you sure that the combiner function in MExtend is stored as syntactically binary?

Yes!

> One way to fix this would be add a utility called asBinaryLam that accepts either a syntactically binary function or a nest of two syntactically unary ones and normalizes to the former.

Part of the goal with this change was to avoid that sort of logic. In fact, in SimpIR, it's not even possible to have a syntactic nest of unary lambdas!